### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference in idx/memory/memory.go:373

### ISSUE THREAD

**Issue by woodsaj** [grafana/metrictank#770]
every node in our worldping cluster failed at the same time today.
```
[Macaron] 2017-11-23 21:27:52: Started GET /metrics/find?from=1511461515&query=*&until=1511472375 for 69.36.243.194
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xa2f7e0]

goroutine 941335569 [running]:
github.com/raintank/metrictank/idx/memory.(*MemoryIdx).find(0xc4207f97a0, 0xb0fe, 0xc4a61976f8, 0x1, 0x0, 0x0, 0xc5206527c0, 0x6e33e6, 0xc4774ef978)
        /home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/idx/memory/memory.go:373 +0x3f0
github.com/raintank/metrictank/idx/memory.(*MemoryIdx).Find(0xc4207f97a0, 0xb0fe, 0xc4a61976f8, 0x1, 0x5a17128b, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/idx/memory/memory.go:276 +0x10b
github.com/raintank/metrictank/api.(*Server).findSeriesLocal(0xc420a1e0a0, 0x102ed20, 0xc424ccce40, 0xb0fe, 0xc471bc7740, 0x1, 0x1, 0x5a17128b, 0x0, 0x0, ...)
        /home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/api/graphite.go:109 +0x2fd
github.com/raintank/metrictank/api.(*Server).findSeries.func1(0xc420a1e0a0, 0x102ed20, 0xc424ccce40, 0xb0fe, 0xc471bc7740, 0x1, 0x1, 0x5a17128b, 0xc471bc7750, 0xc4adc46e80, ...)
        /home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/api/graphite.go:72 +0x9b
created by github.com/raintank/metrictank/api.(*Server).findSeries
        /home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/api/graphite.go:80 +0x4c1
```

running version: 0.7.4-37-gcc0abda

**Comment by Dieterbe** [comment]
looks like the startNode is `nil`.  it's either set via `startNode = tree.Items[""]` or `startNode, ok = tree.Items[branch]`. any chance you still have a record of what the query was @woodsaj ?

**Comment by woodsaj** [comment]
It wasnt my query, all nodes in a cluster crashed and got restarted.  This was the last log entry.

Which is odd, as i thought Macaron was supposed to catch panics and return a 500 instead of killing the process.

**Comment by Dieterbe** [comment]
on api/graphite.go:80 we spawn a separate goroutine in which we run findSeriesLocal. that explains why the stacktrace starts from the `go ...()` line.
when you spawn a separate goroutine and it panics, you can't recover it in the originating goroutine.

**Comment by Dieterbe** [comment]
to solve this I think we can best add some code to `(n *Node) HasChildren()` ; so that if n is nil, we log the path

**Comment by Dieterbe** [comment]
so i looked through the WP index to see if we had any metrics starting with `.` but this does not seem to be the case. i checked via two different commands:

```
$ kubemt exec mt-read00-5000-medium-worldping-a-6f5ddcf7cd-n4ht5 -- sh -c "mt-index-cat --prefix="." cass -timeout 60s -keyspace \$MT_CASSANDRA_IDX_KEYSPACE -hosts \$MT_CASSANDRA_IDX_HOSTS '{{.Id}} {{.OrgId}} {{.Name}} {{.Metric}} {{.Interval}} {{.LastUpdate}} {{.Partition}}\n'"
$
$ kubemt exec mt-read00-5000-medium-worldping-a-6f5ddcf7cd-n4ht5 -- sh -c "mt-index-cat cass -timeout 60s -keyspace \$MT_CASSANDRA_IDX_KEYSPACE -hosts \$MT_CASSANDRA_IDX_HOSTS '={{.Name}} ={{.Metric}} {{.Id}} {{.OrgId}} {{.Interval}} {{.LastUpdate}} {{.Partition}}\n'" > dump.txt
$ grep '=\.' dump.txt
$
```

also:
```
$ grep '\.\.' dump.txt
$
```

so since we don't see any of the common offenders, I suggest we deploy the latest master to WP (which has the extra diagnostics from #812) and if it happens again, we'll have more info.

**Comment by Dieterbe** [comment]
also, seems like all chars in all metrics are sane:
```
$ grep -v '^[-~= \._a-z0-9A-Z]*$' dump.txt
$
```

**Comment by Dieterbe** [comment]
I don't think whatever caused this is fixed, but since #812 was merged: 
1) we won't panic and kill the process anymore
2) we've added detailed error messages in all cases that would have triggered this panic

if we see these errors, we can create a new ticket and look at this again.
